trio/__init__.py

```diff
@@ -13,61 +13,81 @@
+    "TrioInternalError",
+    "RunFinishedError",
+    "WouldBlock",
+    "Cancelled",
+    "ResourceBusyError",
+    "ClosedResourceError",
+    "MultiError",
+    "run",
+    "open_nursery",
+    "open_cancel_scope",
+    "current_effective_deadline",
+    "TASK_STATUS_IGNORED",
+    "current_time",
+    "TaskLocal",
+]
 
-__all__ = []
+from ._version import __version__
 
-from ._toplevel_core_reexports import *
-__all__ += _toplevel_core_reexports.__all__
+from ._core import (
+    TrioInternalError, RunFinishedError, WouldBlock, Cancelled,
+    ResourceBusyError, ClosedResourceError, MultiError, run, open_nursery,
+    open_cancel_scope, current_effective_deadline, TASK_STATUS_IGNORED,
+    current_time, TaskLocal, cancel_shielded_checkpoint
+)
 
-from ._timeouts import *
-__all__ += _timeouts.__all__
+from ._timeouts import (
+    move_on_at, move_on_after, sleep_forever, sleep_until, sleep, fail_at,
+    fail_after, TooSlowError
+)
 
-from ._sync import *
-__all__ += _sync.__all__
+from ._sync import (
+    Event, CapacityLimiter, Semaphore, Lock, StrictFIFOLock, Condition, Queue
+)
 
-from ._threads import *
-__all__ += _threads.__all__
+from ._threads import (
+    run_sync_in_worker_thread, current_default_worker_thread_limiter,
+    BlockingTrioPortal
+)
 
-from ._highlevel_generic import *
-__all__ += _highlevel_generic.__all__
+from ._highlevel_generic import (
+    aclose_forcefully, BrokenStreamError, StapledStream
+)
 
-from ._signals import *
-__all__ += _signals.__all__
+from ._signals import catch_signals
 
-from ._highlevel_socket import *
-__all__ += _highlevel_socket.__all__
+from ._highlevel_socket import SocketStream, SocketListener
 
-from ._file_io import *
-__all__ += _file_io.__all__
+from ._file_io import open_file, wrap_file
 
-from ._path import *
-__all__ += _path.__all__
+from ._path import Path
 
-from ._highlevel_serve_listeners import *
-__all__ += _highlevel_serve_listeners.__all__
+from ._highlevel_serve_listeners import serve_listeners
 
-from ._highlevel_open_tcp_stream import *
-__all__ += _highlevel_open_tcp_stream.__all__
+from ._highlevel_open_tcp_stream import open_tcp_stream
 
-from ._highlevel_open_tcp_listeners import *
-__all__ += _highlevel_open_tcp_listeners.__all__
+from ._highlevel_open_tcp_listeners import open_tcp_listeners, serve_tcp
 
-from ._highlevel_open_unix_stream import *
-__all__ += _highlevel_open_unix_stream.__all__
+from ._highlevel_open_unix_stream import open_unix_socket
 
-from ._highlevel_ssl_helpers import *
-__all__ += _highlevel_ssl_helpers.__all__
+from ._highlevel_ssl_helpers import (
+    open_ssl_over_tcp_stream, open_ssl_over_tcp_listeners, serve_ssl_over_tcp
+)
 
-from ._deprecate import *
-__all__ += _deprecate.__all__
+from ._deprecate import TrioDeprecationWarning
 
 # Imported by default
 from . import hazmat
 from . import socket
 from . import abc
 from . import ssl
 # Not imported by default: testing
+TESTING = False
+if TESTING:
+    from . import testing
```

Comment:
If we do this then it should be `if False: ...` instead of defining a variable `TESTING`. (Remember: every variable we define in this file becomes part of trio's public API!)

I *think* we probably don't need this? From some playing around, it looks like `jedi` is happy to suggest `testing` as a completion even if we don't mention it here:

```python
In [15]: "testing" in [v.name for v in jedi.Script("import trio; trio.").completions()]
Out[15]: True
```

And if I put this into a file and run `pylint` on it, then it doesn't warn on the mention of `trio.testing`, but does warn on the mention of `trio.testingxxx`, so it seems that pylint has also figured out that `trio.testing` is a valid thing:

```python
import trio
print(trio.testing)
print(trio.testingxxx)
```

(These are both with trio v0.6.0, since I happened to have it lying around.)

So there's some puzzle about how pylint actually decides what names are in a namespace: it knows about `testing`, but somehow the way we're querying it currently doesn't...

For now please either make this `if False: ...` or else special-case the name `testing` in our pylint test, and we should look to see if we can figure out pylint's API better...

Reply:
Decided to go for if False: currently to minimise changes.